### DISCUSSION TITLE
Can I have Renovate perform arbitrary but static string replacements?

### DISCUSSION THREAD

**Discussion by reitzig** [renovatebot/renovate#20113] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.48.3

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

8.7

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I want to (ab)use Renovate to update repositories according to central decisions. These can not _really_ be expressed in terms of dependencies, let alone one of the existing data sources.

In the abstract, I want to be able to regex-replace strings without being tied to the execution model behind `regexManagers`.

Concrete use case: We are refactoring Jenkins agents; we want teams to migrate from `label 'old-name'` to `label 'new-name'`. This would tie in well with the workflow teams have been establishing around dependencies.
This _is_ a kind of dependency, after all, if to a set of conventions that change _once_ (as opposed to a list of version strings that keeps extending).

### What you tried so far.

I tried to come up with a kind of "NOP"-datasource that validates and doesn't create a cryptic error during processing.

My best attempt is to use a repology package that _should_ trigger an update, and then use `autoReplaceStringTemplate` to do something entirely different:

```json
{
  "regexManagers": [
    {
      "fileMatch": ["Jenkinsfile*"],
      "matchStrings": ["label\\s*'old-name'"],
      "autoReplaceStringTemplate": "label 'new-name",
      "depNameTemplate": "alpine_edge/curl",
      "datasourceTemplate": "repology",
      "currentValueTemplate": "7.86.0",
      "versioningTemplate": "loose"
    }
  ]
}
```

This is detected "correctly":

```
{
  "regex": [
    {
      "packageFile": "Jenkinsfile",
      "deps": [
        {
          "depName": "alpine_edge/curl",
          "currentValue": "7.86.0",
          "datasource": "repology",
          "versioning": "loose",
          "replaceString": "label 'old-name'",
          "depIndex": 0,
          "updates": [
            {
              "bucket": "non-major",
              "newVersion": "7.87.0-r3",
              "newValue": "7.87.0-r3",
              "newMajor": 7,
              "newMinor": 87,
              "updateType": "minor",
              "branchName": "renovate/alpine_edge-curl-7.x"
            }
          ],
          "warnings": [],
          "currentVersion": "7.86.0",
          "isSingleVersion": true,
          "fixedVersion": "7.86.0"
        }
      ],
      "matchStrings": [
        "label\\s*'old-name'"
      ],
      "depNameTemplate": "alpine_edge/curl",
      "currentValueTemplate": "7.86.0",
      "datasourceTemplate": "repology",
      "versioningTemplate": "loose",
      "autoReplaceStringTemplate": "label 'new-name"
    }
  ]
}
```

The full dry run seems to suggest it would create a PR, but I'm getting a warning I don't understand:

```
DEBUG: Could not extract Jenkinsfile (repository=my-project/my-repo, baseBranch=renovate-testing, branch=renovate/alpine_edge-curl-7.x)
 WARN: Error updating branch: update failure (repository=my-project/my-repo, baseBranch=renovate-testing, branch=renovate/alpine_edge-curl-7.x)
```

Even if it worked it would ... feel dirty? Not least, the PR title would be completely bogus and I don't see a way to overwrite it from config.

Is there a working/cleaner way to do this with the current feature set of Renovate? If not, would it be a reasonable feature to request?

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by reitzig** [comment]
If I were to add this functionality as a "plugin" -- a manager without data sources, methinks? -- is there any guidance on how to go about that? (Not a JS person, me.)

I see that some form of plugin or general data-source mechanism has been requested multiple times, and there doesn't seem to be a straight-forward solution: [#10622](https://github.com/renovatebot/renovate/issues/10622), [#4459](https://github.com/renovatebot/renovate/issues/4459), [#4674](https://github.com/renovatebot/renovate/issues/4674).

Is see the mention of [renovatebot/internal-tools](https://github.com/renovatebot/internal-tools) but have a hard time telling how to adapt it for my needs (again, not a JS person).

**Reply by rarkins** [reply]
A manager whose only purpose is replacements?

Wondering if it's easier/better to add it to the regex manager

**Reply by reitzig** [reply]
Certainly possible! From a user perspective, I felt that the current interface was already complicated enough so I was hesitant to propose an additional, mostly disjoint config path.

That said, something like

```json
{
  "regexManagers": [
    {
      "fileMatch": ["Jenkinsfile*"],
      "matchStrings": ["label\\s*'old-name'"],
      "replacementStringTemplate": "label 'new-name" // might use matching groups
    }
  ]
}
```

that completely skips querying data sources would solve my use case.

Without having looked at the code, I wonder: What would be the update type? How would we determine branch names and PR descriptions? How would I write matching `packageRules` to, say, auto-merge the change? :thinking:

**Reply by rarkins** [reply]
I'm pretty confident it would be best to add this to regex manager, or you'd end up duplicating a lot of code, creating your own bugs, etc.
